storage/s3.go

```diff
@@ -554,11 +581,55 @@ func (s *S3) Put(
 		input.ContentEncoding = aws.String(contentEncoding)
 	}
 
+	// add retry code to the object metadata
+	if s.retryOnNoSuchUploadError > 0 {
+		objectMetaData := make(map[string]*string)
+		num, _ := rand.Int(rand.Reader, big.NewInt(math.MaxInt64))
+		retryCode = num.String()
+		objectMetaData[retryMetaDataKey] = &retryCode
+		input.Metadata = objectMetaData
```

Comment:
If we use other metadata in the future, this line might overwrite it. So it would be nice if we set the default value to `input.Metadata` in the beginning and then set the retry code.